Monochrome bitmap support (bug #18888)

wsu-cb opened the pull request:
Here is a an attempt to provide basic support for monochrome bitmaps, in particular loading from and saving to files.  I also make wxBitmap(const wxImage&, depth 1) generate a monochrome bitmap.  The resulting bitmap is generated by wxQuantize.  I tried to preserve the colors selected by wxQuantize, but was unable to SetDIBColorTable() work.  Therefore, the resulting bitmap is currently always black-and-white.  Does anyone have any insight into what I am doing wrong with SetDIBColorTable()?

I have also included a basic unit test.

wsu-cb committed:
load monochrome .bmp as monochrome wxBitmap

Before this change, loading a monochrome .bmp would result in a
24bpp wxBitmap.  But wxMask requires a monochrome wxBitmap.  So
this patch make loading a monochrome .bmp create a monochrome.
It also allows converting a wxImage with no alpha and no mask to
a monochrome wxBitmap.

wsu-cb committed:
fix some missing spaces

wsu-cb committed:
saving a monochrome bitmap needs to include the color table

wsu-cb committed:
Try (and fail) to include quantized's palette's color table

Can anyone explain why this doesn't work?

wsu-cb committed:
unit test for monochrome bitmaps

Tests:
1) a wxImage can be converted to a monochrome bitmap
2) a monochrome bitmap can be saved to a file
3) a monochrome bitmap can be loaded from a file

vadz commented:
It looks like this broke something, please see unit tests failures on AppVeyor. Could you please check what's going on there?

It also seems that this doesn't work in wxGTK neither and while you don't have to fix this in this PR, of course (although it would be great if you did!), we need to at least avoid breaking the unit tests for it, so we need to check for it and do `WARN("Skipping test known not to work in wxGTK")` inside `#ifdef __WXGTK__`.

vadz commented on src/msw/dib.cpp:
```diff
@@ -598,17 +638,83 @@ bool wxDIB::Create(const wxImage& image, PixelFormat pf)
         return false;
 
+    // convert wxImage's content to monochrome
+    // KLUDGE:  wxScopedPtr doesn't handle T[]
```
But `wxScopedArray` does and should really be used here instead.

vadz commented on src/msw/dib.cpp:
```diff
@@ -598,17 +638,83 @@ bool wxDIB::Create(const wxImage& image, PixelFormat pf)
+
+        // try to use palette's colors in result bitmap
+        MemoryHDC hDC;
+        SelectInHDC(hDC, m_handle);
```
Another bug, I think.

vadz commented on src/msw/dib.cpp:
```diff
@@ -598,17 +638,83 @@ bool wxDIB::Create(const wxImage& image, PixelFormat pf)
+        if ( rc != colors )
+        {
+            wxLogLastError(wxT("SetDIBColorTable"));
+#if 0   // KLUDGE:  SetDIBColorTable always fails for me,
```
What's the last error value? I expect it's something generic and not very useful, but still?

vadz commented on src/msw/dib.cpp:
```diff
@@ -656,10 +762,19 @@ bool wxDIB::Create(const wxImage& image, PixelFormat pf)
-                *dst++ = src[1];
-                *dst++ = src[0];
-                src += 3;
+                if ( bpp != 1 )
```
I haven't measured it, but I think it could be better to move this test out of the loop to allow compiler to handle the loop more efficiently (e.g. auto-vectorize it).

vadz commented on tests/graphics/bitmap.cpp:
```diff
@@ -24,6 +24,7 @@
 #if wxUSE_GRAPHICS_CONTEXT
 #include "wx/graphics.h"
 #endif // wxUSE_GRAPHICS_CONTEXT
+#include "wx/quantize.h"
```
Seems unnecessary.
```suggestion
```

vadz commented on tests/graphics/bitmap.cpp:
```diff
@@ -57,6 +58,23 @@ typedef wxNativePixelData wxNative32PixelData;
+    REQUIRE(color.GetDepth() == 32);
+
+    wxImage imgQuant = color.ConvertToImage();
+    wxBitmap bmpQuant(imgQuant, 1);
```
Shouldn't we have
```
CHECK(bmpQuant.GetDepth() == 1);
```
here?

vadz commented on tests/graphics/bitmap.cpp:
```diff
@@ -57,6 +58,23 @@ typedef wxNativePixelData wxNative32PixelData;
+
+    wxImage imgQuant = color.ConvertToImage();
+    wxBitmap bmpQuant(imgQuant, 1);
+    bmpQuant.SaveFile("quantize.bmp", wxBITMAP_TYPE_BMP);
```
Please use `TestFile` for this file, so it's not left lying around.

It would also be more clear to call it `mono_horse.bmp`, I think.

vadz commented on src/msw/dib.cpp:
```diff
@@ -721,43 +836,55 @@ wxImage wxDIB::ConvertToImage(ConversionFlags flags) const
-            dst[0] = *src++;
-
-            if ( bpp == 32 )
+            if ( bpp != 1 )
```
Same remark as above.

vadz reviewed (changes requested):
Thanks for your changes, but I'm afraid a couple of fixes will be needed, in addition to fixing the previously mentioned test suite problems.

Looking at the code, I'm really not sure why do we need to use not only `SetDIBColorTable()`, but even `GetDIBColorTable()`: does it even make sense to speak about a colour table for a mono bitmap? Do mono bitmaps saved in other applications actually have it?

Thanks!

paulcor commented on tests/graphics/bitmap.cpp:
```diff
@@ -57,6 +58,23 @@ typedef wxNativePixelData wxNative32PixelData;
+    bmpQuant.SaveFile("quantize.bmp", wxBITMAP_TYPE_BMP);
+
+    wxBitmap mono;
+    mono.LoadFile("quantize.bmp", wxBITMAP_TYPE_BMP);
```
I believe only wxMSW can directly load a BMP into a `wxBitmap`, other platforms are going to use `wxImage` to load the BMP and then convert to `wxBitmap`, in which case a depth of 1 won't be preserved.

vadz replied:
In principle, we could have something like `wxIMAGE_OPTION_DEPTH` to remember the depth read from the file on disk and and use its value when converting to `wxBitmap`.

wsu-cb committed:
add missing test validation

wsu-cb committed:
disable monochrome bitmap test in wxGTX since it doesn't work there

wsu-cb commented:
https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/34833364/job/cpgbxtnk5a1xcyg3?fullLog=true

reports that an assert failed at src\msw\bitmap.cpp:1423, but I don't see any call stack report.  Is there any way to at least find out what test was being run when the assert failed?  (When I try to run the full test_gui.exe, it aborts without ever failing on that assert.)





>     On 08/25/2020 11:51 AM VZ <notifications@github.com> wrote:
> 
> 
> 
> 
>     It looks like this broke something, please see unit tests failures on AppVeyor. Could you please check what's going on there?
> 
>     It also seems that this doesn't work in wxGTK neither and while you don't have to fix this in this PR, of course (although it would be great if you did!), we need to at least avoid breaking the unit tests for it, so we need to check for it and do WARN("Skipping test known not to work in wxGTK") inside #ifdef __WXGTK__.
> 
>     —
>     You are receiving this because you authored the thread.
>     Reply to this email directly, view it on GitHub https://github.com/wxWidgets/wxWidgets/pull/2032#issuecomment-680110498 , or unsubscribe https://github.com/notifications/unsubscribe-auth/AQVEPLVCFFZGH7Z5ECCQTELSCPMXZANCNFSM4QKGSFEA .
>

wsu-cb committed:
use anonymous namespace instead of static functions

wsu-cb committed:
fix error that immediately unselected bitmap from DC

wsu-cb committed:
make Save() function easier to read (hopefully)

wsu-cb committed:
Use wxCHECK_MSG instead of wxLogError for (monochrome + alpha) error

vadz commented:
> https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/34833364/job/cpgbxtnk5a1xcyg3?fullLog=true reports that an assert failed at src\msw\bitmap.cpp:1423, but I don't see any call stack report. Is there any way to at least find out what test was being run when the assert failed? 

It's supposed to say during which test this happened, but clearly this didn't work in this case :-(

> (When I try to run the full test_gui.exe, it aborts without ever failing on that assert.)

It's definitely not supposed to abort, does it happen for you even before your changes (i.e. with master)? Where does it happen (you should be able to see it if you run it under debugger or if you have a [post-mortem debugger](https://docs.microsoft.com/en-us/windows/win32/debug/configuring-automatic-debugging) configured)?

wsu-cb committed:
replace Cleaner hack with wxScopedArray

wsu-cb committed:
remove unnecessary #include

wsu-cb committed:
avoid performing bpp != 1 test on every loop iteration

wsu-cb committed:
SetDIBColorTable() problem solved, so get rid of experimental code

wsu-cb committed:
remove test-generated file when test exits

wsu-cb committed:
avoid performing bpp != 1 test on every loop iteration

wsu-cb commented:
I have pushed my updates to address these issues, but there doesn't seem to be much point to another PR until the bitmap.cpp assert failure.



>     On 08/25/2020 12:09 PM VZ <notifications@github.com> wrote:
> 
> 
> 
> 
>     @vadz requested changes on this pull request.
> 
>     Thanks for your changes, but I'm afraid a couple of fixes will be needed, in addition to fixing the previously mentioned test suite problems.
> 
>     Looking at the code, I'm really not sure why do we need to use not only SetDIBColorTable(), but even GetDIBColorTable(): does it even make sense to speak about a colour table for a mono bitmap? Do mono bitmaps saved in other applications actually have it?
> 
>     Thanks!
> 

The GetDIBColorTable was just demonstrating some things I was doing to try to get the SetDIBColorTable() to work (see below).  When Microsoft Visual Studio 2019 converts horse.bmp to two colors, neither of them are black or white.  Full disclosure:  the resulting .bmp is weird.  The file is approximately 2.5 times bigger than the original!  However, the header says it's a monochrome .bmp with the expected size, even though the actual file is huge.  Bizarre.  It looks like even Microsoft doesn't deal well with depth < 24 nowadays.  If you look at the results of the conversion and don't like them, I can go back to generating pure black and pure white as the color table, which is what the bad SetDIBColorTable() resulted in.




>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476555534 :
> 
>     > @@ -71,6 +72,23 @@ static inline bool GetDIBSection(HBITMAP hbmp, DIBSECTION *ds)
> 
>                      ds->dsBm.bmBits;
> 
>      }
> 
>      
> 
>     +// for monochrome bitmaps, need bit twiddling functions to get at pixels
> 
>     +static inline bool MonochromeLineReadBit(const unsigned char* srcLineStart, int index)
> 
> 
>     Slight nitpick:
> 
>     ⬇️ Suggested change
> 
>     -static inline bool MonochromeLineReadBit(const unsigned char* srcLineStart, int index)
> 
>     +namespace {
> 
>     +
> 
>     +inline bool MonochromeLineReadBit(const unsigned char* srcLineStart, int index)
> 
> 
>     (and remove static for the other function too and add } // anonymous namespace below then). I.e. using anonymous namespace for local functions is slightly preferable to using static.
> 


I prefer anonymous namespace as well, but the original file used static functions.  I have now committed anonymous namespace.


>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476556322 :
> 
>     > @@ -249,15 +267,34 @@ bool wxDIB::Save(const wxString& filename)
> 
>                  const size_t sizeHdr = ds.dsBmih.biSize;
> 
>                  const size_t sizeImage = ds.dsBmih.biSizeImage;
> 
>      
> 
>     +            // provide extra space so we can verify that DIB's
> 
>     +            // color table is size 2
> 
>     +            RGBQUAD monoBmiColors[3];
> 
>     +            if ( ds.dsBmih.biBitCount == 1 )
> 
>     +            {
> 
>     +                memset(monoBmiColors, 128, sizeof(monoBmiColors));
> 
> 
>     Could you please add a comment explaining where does this 128 come from?
> 

Again, this was experimenting to try to understand the SetDIBColorTable() problem.  It will go away now.




>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476556622 :
> 
>     >              bmpHdr.bfType = 0x4d42;    // 'BM' in little endian
> 
>     -            bmpHdr.bfOffBits = sizeof(BITMAPFILEHEADER) + ds.dsBmih.biSize;
> 
>     +            bmpHdr.bfOffBits = sizeof(BITMAPFILEHEADER);
> 
>     +            bmpHdr.bfOffBits += ds.dsBmih.biSize;
> 
>     +            bmpHdr.bfOffBits += (ds.dsBmih.biBitCount == 1 ? sizeof(monoBmiColors) : 0);
> 
> 
>     ⬇️ Suggested change
> 
>     -            bmpHdr.bfOffBits += (ds.dsBmih.biBitCount == 1 ? sizeof(monoBmiColors) : 0);
> 
>     +            bmpHdr.bfOffBits += ds.dsBmih.biBitCount == 1 ? sizeof(monoBmiColors) : 0;
> 

I did something different here.  See what you think of my new version.




> 
> 
>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476558272 :
> 
>     > @@ -249,15 +267,34 @@ bool wxDIB::Save(const wxString& filename)
> 
>                  const size_t sizeHdr = ds.dsBmih.biSize;
> 
>                  const size_t sizeImage = ds.dsBmih.biSizeImage;
> 
>      
> 
>     +            // provide extra space so we can verify that DIB's
> 
>     +            // color table is size 2
> 
>     +            RGBQUAD monoBmiColors[3];
> 
>     +            if ( ds.dsBmih.biBitCount == 1 )
> 
>     +            {
> 
>     +                memset(monoBmiColors, 128, sizeof(monoBmiColors));
> 
>     +                MemoryHDC dc;
> 
>     +                SelectInHDC(dc, m_handle);
> 
> 
>     This looks like a mistake and should almost certainly be
> 
>     ⬇️ Suggested change
> 
>     -                SelectInHDC(dc, m_handle);
> 
>     +                SelectInHDC selectInDC(dc, m_handle);
> 
> 
>     as otherwise you just select and deselect it immediately, without any real effect.
> 



You are so right!


>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476558782 :
> 
>     >              bmpHdr.bfSize = bmpHdr.bfOffBits + sizeImage;
> 
>      
> 
>                  // first write the file header, then the bitmap header and finally the
> 
>                  // bitmap data itself
> 
>                  ok = file.Write(&bmpHdr, sizeof(bmpHdr)) == sizeof(bmpHdr) &&
> 
>                          file.Write(&ds.dsBmih, sizeHdr) == sizeHdr &&
> 
>     -                        file.Write(ds.dsBm.bmBits, sizeImage) == sizeImage;
> 
>     +                        (ds.dsBmih.biBitCount != 1 || file.Write(monoBmiColors, sizeof(monoBmiColors)) == sizeof(monoBmiColors)) &&
> 
> 
>     Would be nice to make this line shorter, maybe by adding sizeCols variable to avoid writing long sizeof(monoBmiColors) multiple times.
> 

Please see how I rewrote this section.


>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476559176 :
> 
>     >      const bool hasAlpha = image.HasAlpha();
> 
>     -    const int bpp = hasAlpha ? 32 : 24;
> 
>     +    // monochrome DIBs can't express alpha
> 
>     +    if ( hasAlpha && depth == 1 ) {
> 
> 
>     Just a style nitpick:
> 
>     ⬇️ Suggested change
> 
>     -    if ( hasAlpha && depth == 1 ) {
> 
>     +    if ( hasAlpha && depth == 1 )
> 
>     +    {
> 

> 
> 
>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476560150 :
> 
>     >      const bool hasAlpha = image.HasAlpha();
> 
>     -    const int bpp = hasAlpha ? 32 : 24;
> 
>     +    // monochrome DIBs can't express alpha
> 
>     +    if ( hasAlpha && depth == 1 ) {
> 
>     +        wxLogError( _("can't convert image with alpha to monochrome") );
> 
> 
>     Not really sure if we want to have this as a user-visible message. Maybe this should be an assert instead, i.e. replace the entire if with
> 
>     wxCHECK_MSG( !hasAlpha || depth != 1, false, "alpha not supported in monochrome bitmaps" );
> 
> 
>     ?
> 

OK.


>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476560526 :
> 
>     >  
> 
>          if ( !Create(w, h, bpp) )
> 
>              return false;
> 
>      
> 
>     +    // convert wxImage's content to monochrome
> 
>     +    // KLUDGE:  wxScopedPtr doesn't handle T[]
> 
> 
>     But wxScopedArray does and should really be used here instead.
> 

Good to know.



>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476560833 :
> 
>     > +        wxPalette* tempPalette;
> 
>     +        if ( !wxQuantize::Quantize(
> 
>     +            image,
> 
>     +            quantized,
> 
>     +            &tempPalette,
> 
>     +            2,
> 
>     +            &eightBitData,
> 
>     +            wxQUANTIZE_RETURN_8BIT_DATA) )
> 
>     +        {
> 
>     +            return false;
> 
>     +        }
> 
>     +        wxScopedPtr<wxPalette> palette(tempPalette);
> 
>     +
> 
>     +        // try to use palette's colors in result bitmap
> 
>     +        MemoryHDC hDC;
> 
>     +        SelectInHDC(hDC, m_handle);
> 
> 
>     Another bug, I think.
> 


Ding!  Ding!  Ding!
That's the problem with SetDIBColorTable()!  The bitmap isn't in SetDIBColorTable()'s DC.



>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476561376 :
> 
>     > +        for ( UINT i = 0; i < colors; ++i )
> 
>     +        {
> 
>     +            if ( !palette->GetRGB(i,
> 
>     +                                    &colorTable[i].rgbRed,
> 
>     +                                    &colorTable[i].rgbGreen,
> 
>     +                                    &colorTable[i].rgbBlue) )
> 
>     +            {
> 
>     +                return false;
> 
>     +            }
> 
>     +            colorTable[i].rgbReserved = 0;
> 
>     +        }
> 
>     +        UINT rc = SetDIBColorTable(hDC, 0, colors, colorTable);
> 
>     +        if ( rc != colors )
> 
>     +        {
> 
>     +            wxLogLastError(wxT("SetDIBColorTable"));
> 
>     +#if 0   // KLUDGE:  SetDIBColorTable always fails for me,
> 
> 
>     What's the last error value? I expect it's something generic and not very useful, but still?
> 

Fixed, so no longer relevant.


>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476561874 :
> 
>     > @@ -656,10 +762,19 @@ bool wxDIB::Create(const wxImage& image, PixelFormat pf)
> 
>              {
> 
>                  for ( int x = 0; x < w; x++ )
> 
>                  {
> 
>     -                *dst++ = src[2];
> 
>     -                *dst++ = src[1];
> 
>     -                *dst++ = src[0];
> 
>     -                src += 3;
> 
>     +                if ( bpp != 1 )
> 
> 
>     I haven't measured it, but I think it could be better to move this test out of the loop to allow compiler to handle the loop more efficiently (e.g. auto-vectorize it).
> 

Good point.  Even if it doesn't vectorize, doing the bpp != 1 test on every loop iteration must be a waste.
I lifted the bpp != 1 test out one loop level; did you mean to lift both levels?  That would mean a bit more code replication in the if-body and else-body.


>     ---------------------------------------------
> 
>     In tests/graphics/bitmap.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476562758 :
> 
>     > @@ -24,6 +24,7 @@
> 
>      #if wxUSE_GRAPHICS_CONTEXT
> 
>      #include "wx/graphics.h"
> 
>      #endif // wxUSE_GRAPHICS_CONTEXT
> 
>     +#include "wx/quantize.h"
> 
> 
>     Seems unnecessary.
> 
>     ⬇️ Suggested change
> 
>     -#include "wx/quantize.h"
> 

Done.


> 
> 
>     ---------------------------------------------
> 
>     In tests/graphics/bitmap.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476563383 :
> 
>     > @@ -57,6 +58,23 @@ typedef wxNativePixelData wxNative32PixelData;
> 
>      // tests
> 
>      // ----------------------------------------------------------------------------
> 
>      
> 
>     +TEST_CASE("BitmapTestCase::Monochrome", "[bitmap][monochrome]")
> 
>     +{
> 
>     +    wxBitmap color;
> 
>     +    color.LoadFile("horse.bmp", wxBITMAP_TYPE_BMP);
> 
>     +    REQUIRE(color.IsOk());
> 
>     +    REQUIRE(color.GetDepth() == 32);
> 
>     +
> 
>     +    wxImage imgQuant = color.ConvertToImage();
> 
>     +    wxBitmap bmpQuant(imgQuant, 1);
> 
> 
>     Shouldn't we have
> 
>     CHECK(bmpQuant.GetDepth() == 1);
> 
> 
>     here?
> 

I already had that in my local working copy.  I just hadn't committed/pushed it yet.


>     ---------------------------------------------
> 
>     In tests/graphics/bitmap.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476563857 :
> 
>     > @@ -57,6 +58,23 @@ typedef wxNativePixelData wxNative32PixelData;
> 
>      // tests
> 
>      // ----------------------------------------------------------------------------
> 
>      
> 
>     +TEST_CASE("BitmapTestCase::Monochrome", "[bitmap][monochrome]")
> 
>     +{
> 
>     +    wxBitmap color;
> 
>     +    color.LoadFile("horse.bmp", wxBITMAP_TYPE_BMP);
> 
>     +    REQUIRE(color.IsOk());
> 
>     +    REQUIRE(color.GetDepth() == 32);
> 
>     +
> 
>     +    wxImage imgQuant = color.ConvertToImage();
> 
>     +    wxBitmap bmpQuant(imgQuant, 1);
> 
>     +    bmpQuant.SaveFile("quantize.bmp", wxBITMAP_TYPE_BMP);
> 
> 
>     Please use TestFile for this file, so it's not left lying around.
> 
>     It would also be more clear to call it mono_horse.bmp, I think.
> 

After puzzling over this, I think you mean TempFile, not TestFile.  Assuming that is what you mean, done.



>     ---------------------------------------------
> 
>     In src/msw/dib.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r476564442 :
> 
>     > @@ -721,43 +836,55 @@ wxImage wxDIB::ConvertToImage(ConversionFlags flags) const
> 
>              const unsigned char *src = srcLineStart;
> 
>              for ( int x = 0; x < w; x++ )
> 
>              {
> 
>     -            dst[2] = *src++;
> 
>     -            dst[1] = *src++;
> 
>     -            dst[0] = *src++;
> 
>     -
> 
>     -            if ( bpp == 32 )
> 
>     +            if ( bpp != 1 )
> 
> 
>     Same remark as above.
> 

Same as above :-)


> 
>     —You are receiving this because you authored the thread.
>     Reply to this email directly, view it on GitHub https://github.com/wxWidgets/wxWidgets/pull/2032#pullrequestreview-474620440 , or unsubscribe https://github.com/notifications/unsubscribe-auth/AQVEPLTGENI6ZK4RQRHYZNDSCPO2DANCNFSM4QKGSFEA .
>

wsu-cb commented:
>     On 08/26/2020 7:56 AM VZ <notifications@github.com> wrote:
> 
> 
> 
> 
>     @vadz commented on this pull request.
> 
> 
>     ---------------------------------------------
> 
>     In tests/graphics/bitmap.cpp https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r477242970 :
> 
>     > @@ -57,6 +58,23 @@ typedef wxNativePixelData wxNative32PixelData;
>      // tests
>      // ----------------------------------------------------------------------------
>      
>     +TEST_CASE("BitmapTestCase::Monochrome", "[bitmap][monochrome]")
>     +{
>     +    wxBitmap color;
>     +    color.LoadFile("horse.bmp", wxBITMAP_TYPE_BMP);
>     +    REQUIRE(color.IsOk());
>     +    REQUIRE(color.GetDepth() == 32);
>     +
>     +    wxImage imgQuant = color.ConvertToImage();
>     +    wxBitmap bmpQuant(imgQuant, 1);
>     +    bmpQuant.SaveFile("quantize.bmp", wxBITMAP_TYPE_BMP);
>     +
>     +    wxBitmap mono;
>     +    mono.LoadFile("quantize.bmp", wxBITMAP_TYPE_BMP);
> 
>     In principle, we could have something like wxIMAGE_OPTION_DEPTH to remember the depth read from the file on disk and and use its value when converting to wxBitmap.
> 


Do you mean a defaulted argument on wxBitmap::LoadFile()?  Frankly, I don't see much point in making that an option.  I think it should always have been doing that, and the current state is a bug, not a compatibility-preserved requirement.  I realize that the other 2/4/8 bpp files are still buggy, but it still seems like a bug to me.  I'm just not ambitious enough to fix them since I'm not using them.


> 
>     —You are receiving this because you authored the thread.
>     Reply to this email directly, view it on GitHub https://github.com/wxWidgets/wxWidgets/pull/2032#discussion_r477242970 , or unsubscribe https://github.com/notifications/unsubscribe-auth/AQVEPLUL7WLHRAHJ6KLRKBLSCTZ5LANCNFSM4QKGSFEA .
>

wsu-cb committed:
Merge branch 'master' into monochrome-bitmap

vadz commented:
Thanks for the updates! Unfortunately there are still plenty of CI failures, I haven't had time to look at them yet, but they certainly need to be fixed before this can be merged. For now just a couple of quick remarks to address some parts of your reply:

> When Microsoft Visual Studio 2019 converts horse.bmp to two colors, neither of them are black or white.  Full disclosure:  the resulting .bmp is weird.  The file is approximately 2.5 times bigger than the original!  However, the header says it's a monochrome .bmp with the expected size, even though the actual file is huge.  Bizarre.  It looks like even Microsoft doesn't deal well with depth < 24 nowadays.  If you look at the results of the conversion and don't like them, I can go back to generating pure black and pure white as the color table, which is what the bad SetDIBColorTable() resulted in.

I'm pretty sure Microsoft still supports mono bitmaps, they're still used as masks for cursors and icons (in fact I think they still support 4/8/16bpp bitmaps too, and at least only a few years ago RDP used to downgrade to 16bpp almost by default, so this depth was still relatively widely used). And it would be best to create the same bitmaps as the other tools create. I don't think the colours in the mono bitmap matter at all, only their indices do.

> After puzzling over this, I think you mean TempFile, not TestFile.  Assuming that is what you mean, done.

No, I did write what I wanted for once :-) There is `TestFile` class in `tests/testfile.h`, it just removes the file it creates in its dtor, thus ensuring that it's not left on disk even if the test fails.

> Do you mean a defaulted argument on wxBitmap::LoadFile()?

No, not at all. The problem is that `wxImage` needs to somehow remember that it loaded a mono bitmap, but it doesn't store depth (it's always 24/32bpp), so we can only use its options mechanism (see its `SetOption()/GetOption()/GetOptionInt()` methods) for this. I do agree with you that the fact that the depth is just lost currently is a bug, not a feature.

But implementing this would be best done in a separate PR, once this one is merged. Right now you just really need to skip these tests on non-MSW platforms, to avoid breaking the test suite. TIA!

wsu-cb commented:
>     On 08/27/2020 7:06 AM VZ <notifications@github.com> wrote:
> 
> 
> 
> 
>     Thanks for the updates! Unfortunately there are still plenty of CI failures, I haven't had time to look at them yet, but they certainly need to be fixed before this can be merged. For now just a couple of quick remarks to address some parts of your reply:
> 
>         > > 
> >         When Microsoft Visual Studio 2019 converts horse.bmp to two colors, neither of them are black or white. Full disclosure: the resulting .bmp is weird. The file is approximately 2.5 times bigger than the original! However, the header says it's a monochrome .bmp with the expected size, even though the actual file is huge. Bizarre. It looks like even Microsoft doesn't deal well with depth < 24 nowadays. If you look at the results of the conversion and don't like them, I can go back to generating pure black and pure white as the color table, which is what the bad SetDIBColorTable() resulted in.
> > 
> >     > 
>     I'm pretty sure Microsoft still supports mono bitmaps, they're still used as masks for cursors and icons (in fact I think they still support 4/8/16bpp bitmaps too, and at least only a few years ago RDP used to downgrade to 16bpp almost by default, so this depth was still relatively widely used). And it would be best to create the same bitmaps as the other tools create. I don't think the colours in the mono bitmap matter at all, only their indices do.
> 

Still, that VS2019 takes 32bpp .bmp, converts it to 1bpp, and produces a larger file is ... disappointing.

At any rate, I take it that you are suggesting I go back to generating pure black/white wxBitmap rather than using wxQuantize's chosen gray values.


> 
>         > > 
> >         After puzzling over this, I think you mean TempFile, not TestFile. Assuming that is what you mean, done.
> > 
> >     > 
>     No, I did write what I wanted for once :-) There is TestFile class in tests/testfile.h, it just removes the file it creates in its dtor, thus ensuring that it's not left on disk even if the test fails.
> 

Are you absolutely sure you mean TestFile?  Both TempFile and TestFile are in testfile.h, and both remove their file in the dtor.  However, TestFile gives the caller no choice in the filename, where TempFile does.  You suggested using name "mono_horse.bmp", which leads me to think you actually want to set the name, which leads me to think you mean TempFile, not TestFile.


> 
>         > > 
> >         Do you mean a defaulted argument on wxBitmap::LoadFile()?
> > 
> >     > 
>     No, not at all. The problem is that wxImage needs to somehow remember that it loaded a mono bitmap, but it doesn't store depth (it's always 24/32bpp), so we can only use its options mechanism (see its SetOption()/GetOption()/GetOptionInt() methods) for this. I do agree with you that the fact that the depth is just lost currently is a bug, not a feature.
> 

Let me see if this is what you are saying:  wxBitmap::LoadFile() should (as I am trying to achieve) create a mono bitmap from a mono .bmp.  wxImage::LoadFile() will (as ever) create a 24/32bpp wxImage, but store in its "userdata" (option) the original depth.  Then wxBitmap::wxBitmap(const wxImage&, int depth = wxBITMAP_SCREEN_DEPTH) will work as currently, but a new wxBitmap::wxBitmap(const wxImage&, int depth = wxBITMAP_ORIGINAL_DEPTH) (depth -2) will result in a wxBitmap of 1 bpp for mono .bmp.  The 2/4/8bpp will necessarily also act as currently since I am not trying to implement them.

> 
>     But implementing this would be best done in a separate PR, once this one is merged. Right now you just really need to skip these tests on non-MSW platforms, to avoid breaking the test suite. TIA!
> 

Of course.  But is this test failing on all non-wxMSW targets, or just wxGTK?


> 
>     —You are receiving this because you authored the thread.
>     Reply to this email directly, view it on GitHub https://github.com/wxWidgets/wxWidgets/pull/2032#issuecomment-681880381 , or unsubscribe https://github.com/notifications/unsubscribe-auth/AQVEPLXRVX72SNGHHENPA5LSCY42NANCNFSM4QKGSFEA .
>

wsu-cb commented:
Don't spend time on the test failures yet.  I'm on the trail of a potential explanation...

>     On 08/27/2020 7:06 AM VZ <notifications@github.com> wrote:
> 
> 
> 
> 
>     Thanks for the updates! Unfortunately there are still plenty of CI failures, I haven't had time to look at them yet, but they certainly need to be fixed before this can be merged. For now just a couple of quick remarks to address some parts of your reply:
> 
>         > > 
> >         When Microsoft Visual Studio 2019 converts horse.bmp to two colors, neither of them are black or white. Full disclosure: the resulting .bmp is weird. The file is approximately 2.5 times bigger than the original! However, the header says it's a monochrome .bmp with the expected size, even though the actual file is huge. Bizarre. It looks like even Microsoft doesn't deal well with depth < 24 nowadays. If you look at the results of the conversion and don't like them, I can go back to generating pure black and pure white as the color table, which is what the bad SetDIBColorTable() resulted in.
> > 
> >     > 
>     I'm pretty sure Microsoft still supports mono bitmaps, they're still used as masks for cursors and icons (in fact I think they still support 4/8/16bpp bitmaps too, and at least only a few years ago RDP used to downgrade to 16bpp almost by default, so this depth was still relatively widely used). And it would be best to create the same bitmaps as the other tools create. I don't think the colours in the mono bitmap matter at all, only their indices do.
> 
>         > > 
> >         After puzzling over this, I think you mean TempFile, not TestFile. Assuming that is what you mean, done.
> > 
> >     > 
>     No, I did write what I wanted for once :-) There is TestFile class in tests/testfile.h, it just removes the file it creates in its dtor, thus ensuring that it's not left on disk even if the test fails.
> 
>         > > 
> >         Do you mean a defaulted argument on wxBitmap::LoadFile()?
> > 
> >     > 
>     No, not at all. The problem is that wxImage needs to somehow remember that it loaded a mono bitmap, but it doesn't store depth (it's always 24/32bpp), so we can only use its options mechanism (see its SetOption()/GetOption()/GetOptionInt() methods) for this. I do agree with you that the fact that the depth is just lost currently is a bug, not a feature.
> 
>     But implementing this would be best done in a separate PR, once this one is merged. Right now you just really need to skip these tests on non-MSW platforms, to avoid breaking the test suite. TIA!
> 
>     —
>     You are receiving this because you authored the thread.
>     Reply to this email directly, view it on GitHub https://github.com/wxWidgets/wxWidgets/pull/2032#issuecomment-681880381 , or unsubscribe https://github.com/notifications/unsubscribe-auth/AQVEPLXRVX72SNGHHENPA5LSCY42NANCNFSM4QKGSFEA .
>

wsu-cb committed:
clarify which variable is source depth and which is destination depth

wsu-cb committed:
fix missing pointer update

wsu-cb commented:
I don't know if this is the problem AppVeyor was reporting, but if I run just the test BitmapTestCase::DrawAlphaWithMask,  I get:

Test program for wxWidgets GUI features
build: 3.1.5 (wchar_t,Visual C++ 1900,wx containers,compatible with 3.0)
running under Windows 10 (build 18363), 64-bit edition as wsu, locale is C
wxWidgets assert: "Assert failure" failed at G:\cyberboard\cb4.test\wxWidgets\src\msw\bitmap.cpp:1423 in wxBitmap::GetRawData with message 'incorrect bitmap type in wxBitmap::GetRawData()'
and another wxWidgets assert: !m_dib failed at G:\cyberboard\cb4.test\wxWidgets\src\msw\bitmap.cpp:273 in wxBitmapRefData::Free with message 'forgot to call wxBitmap::UngetRawData()!' while handling it.
G:\cyberboard\cb4.test\wxWidgets\tests\vc_x64_mswud\test_gui.exe (process 19892) exited with code -1.
To automatically close the console when debugging stops, enable Tools->Options->Debugging->Automatically close the console when debugging stops.
Press any key to close this window . . .


I think I understand what is happening here, but it's somewhat complicated. https://docs.wxwidgets.org/trunk/classwx_pixel_data.html

https://docs.wxwidgets.org/trunk/classwx_pixel_data.html says that it is required that a wxBitmap be created at depth 24 or 32 to use wxPixelData (on wxMSW).  However test BitmapTestCase::DrawAlphaWithMask makes calls that eventually reach wxBitmap::MSWBlendMaskWithAlpha().  wxBitmap::MSWBlendMaskWithAlpha() has in it:

wxBitmap bmpMask = GetMask()->GetBitmap();
wxNativePixelData maskData(bmpMask);

As we have noted before, a mask's bitmap is expected to be monochrome, and yet wxWidgets's own code is trying to use wxNativePixelData on a mask's bitmap. And before my changes, this worked, because wxBitmap::GetRawData() internally converted the monochrome bitmap to 24bpp using the fact that wxDIB::Create(int width, int height, int depth) actually created a 24bpp DIB despite the argument depth = 1!  However, with my changes, which I think make wxDIB::Create(int width, int height, int depth) behavior more "correct", wxNativePixelData fails on a monochrome bitmap since the result of wxDIB::Create(depth=1) is monochrome, not 24bpp, and wxNativePixelData can't work on monochrome bitmaps..

I think, with some work, I can make wxBitmap::GetRawData() pass through the intended depth=24 to wxDIB::Create(), but to me, there is the question of whether wxBitmap::GetRawData() should try to do this when the documentation says wxPixelData can only be used on 24/32bpp wxBitmap.






>     On 08/28/2020 2:57 PM Bill Su <billsu@comcast.net> wrote:
> 
> 
>     Don't spend time on the test failures yet.  I'm on the trail of a potential explanation...
> 
>         > >         On 08/27/2020 7:06 AM VZ <notifications@github.com> wrote:
> > 
> > 
> > 
> > 
> >         Thanks for the updates! Unfortunately there are still plenty of CI failures, I haven't had time to look at them yet, but they certainly need to be fixed before this can be merged. For now just a couple of quick remarks to address some parts of your reply:
> > 
> >             > > > 
> > >             When Microsoft Visual Studio 2019 converts horse.bmp to two colors, neither of them are black or white. Full disclosure: the resulting .bmp is weird. The file is approximately 2.5 times bigger than the original! However, the header says it's a monochrome .bmp with the expected size, even though the actual file is huge. Bizarre. It looks like even Microsoft doesn't deal well with depth < 24 nowadays. If you look at the results of the conversion and don't like them, I can go back to generating pure black and pure white as the color table, which is what the bad SetDIBColorTable() resulted in.
> > > 
> > >         > > 
> >         I'm pretty sure Microsoft still supports mono bitmaps, they're still used as masks for cursors and icons (in fact I think they still support 4/8/16bpp bitmaps too, and at least only a few years ago RDP used to downgrade to 16bpp almost by default, so this depth was still relatively widely used). And it would be best to create the same bitmaps as the other tools create. I don't think the colours in the mono bitmap matter at all, only their indices do.
> > 
> >             > > > 
> > >             After puzzling over this, I think you mean TempFile, not TestFile. Assuming that is what you mean, done.
> > > 
> > >         > > 
> >         No, I did write what I wanted for once :-) There is TestFile class in tests/testfile.h, it just removes the file it creates in its dtor, thus ensuring that it's not left on disk even if the test fails.
> > 
> >             > > > 
> > >             Do you mean a defaulted argument on wxBitmap::LoadFile()?
> > > 
> > >         > > 
> >         No, not at all. The problem is that wxImage needs to somehow remember that it loaded a mono bitmap, but it doesn't store depth (it's always 24/32bpp), so we can only use its options mechanism (see its SetOption()/GetOption()/GetOptionInt() methods) for this. I do agree with you that the fact that the depth is just lost currently is a bug, not a feature.
> > 
> >         But implementing this would be best done in a separate PR, once this one is merged. Right now you just really need to skip these tests on non-MSW platforms, to avoid breaking the test suite. TIA!
> > 
> >         —
> >         You are receiving this because you authored the thread.
> >         Reply to this email directly, view it on GitHub https://github.com/wxWidgets/wxWidgets/pull/2032#issuecomment-681880381 , or unsubscribe https://github.com/notifications/unsubscribe-auth/AQVEPLXRVX72SNGHHENPA5LSCY42NANCNFSM4QKGSFEA .
> > 
> >     >

wsu-cb committed:
see if adding ; solves compile problem on wxGTK

vadz commented:
Thanks for explaining the problem. `wxBitmap::MSWBlendMaskWithAlpha()` could probably be rewritten to avoid raw pixel access, but a worse problem is that I think it's quite likely that people used `wxNativePixelData` with "mono" bitmaps (that actually were 24bpp all this time) in their own code.

So if we somehow could make `wxNativePixelData` work with them, this would be the best, because the most compatibility-preserving, fix. Otherwise we'd need to at least very clearly document this in the "incompatible changes" section of the changelog. And rewrite `wxBitmap::MSWBlendMaskWithAlpha()` without using `wxNativePixelData` for the mask, of course...

wsu-cb commented:
> So if we somehow could make `wxNativePixelData` work with them, this would be the best, because the most compatibility-preserving, fix. Otherwise we'd need to at least very clearly document this in the "incompatible changes" section of the changelog. And rewrite `wxBitmap::MSWBlendMaskWithAlpha()` without using `wxNativePixelData` for the mask, of course...


I think I see three broad approaches to making wxNativePixelData work with mono wxBitmap.  Which of the following do you prefer?  (Note:  I have attempted none of these, so I could be wrong about them being feasible.)

1)  Make wxNativePixelData pay attention to the actual bpp of wxBitmap, and use different logic for 24bpp and 1bpp.  This has the clear disadvantage that virtually every operation will require if-tests checking actual bpp and more memory to store, badly impacting the intended high efficiency of NativePixelData.

2)  Create a new wxNativeMonoPixelData.  As efficient as possible, but not fully source code compatible.  This is actually my preference.

3)  Use the trick from before this PR of having wxBitmap::GetRawData() use wxDIB::Create(depth=24) to create a temporary bitmap and then convert it back when done iterating.  However, it is somewhat misleading to imply that wxNativePixelData can separately read and modify Red, Green, and Blue channels of a mono bitmap.  The temporary bitmap will have to be wxQuantized down to 1bpp when converting back to mono.  I suspect this is your preference.

wsu-cb commented:
Now that I think about it, do the non-wxMSW versions of wxWidgets all have only a single supported native bitmap depth (non-alpha), allowing wxNativePixelData to actually be valid?  Or do they do the same temporary bitmap conversion trick?

vadz commented:
Yes, `wxNativeMonoPixelData` would probably be the best. It's not backwards-compatible but I don't see what else could we do, to be honest. That is, other than continuing to create 24bpp bitmaps when `depth==1`, but this doesn't seem to be optimal neither :-(

wsu-cb committed:
wxPixelData support for monochrome wxBitmap

The internal format of monochrome and 24/32bpp bitmaps are too
different for wxNativePixelData to efficiently handle both, so
create wxNativeMonoPixelData to handle monochrome bitmap.  Then
use wxNativeMonoPixelData as needed.

wsu-cb committed:
fix a missing __WXMSW__  guard

wsu-cb committed:
see if moving explicit specialization outside containing class definition satisfies mingw

wsu-cb committed:
try to fix mingw friend syntax error

wsu-cb committed:
wxNativeMonoPixelData is only implemennted in wxMSW

wsu-cb committed:
fix missing ;

wsu-cb committed:
restore non-wxMSW version of wxNativePixelData tests

wsu-cb commented:
> Yes, `wxNativeMonoPixelData` would probably be the best. It's not backwards-compatible but I don't see what else could we do, to be honest. That is, other than continuing to create 24bpp bitmaps when `depth==1`, but this doesn't seem to be optimal neither :-(

I have finally managed to provide wxNativeMonoPixelData for wxMSW and pass the automated checks, so I am ready for your next review.

Is a message of this sort appropriate, or do you automatically get notices of updates to Pull Requests?

vadz commented:
> Is a message of this sort appropriate, or do you automatically get notices of updates to Pull Requests?

Yes, thank you. I do get updates for the PRs I had already commented on, but it's not always obvious if it's the last planned change or if there are more to come, so it's useful to know when it's done, don't hesitate to ping me if I don't react after a couple of days. I'll look at this one a.s.a.p., although probably not today.

wsu-cb commented:
> ... don't hesitate to ping me if I don't react after a couple of days. I'll look at this one a.s.a.p., although probably not today.

Ping

vadz reviewed (dismissed):
Sorry, I've started reviewing this from top to bottom, and actually it was looking pretty nice, but then I've realized I should look at the test changes first and there it became clear that this is really a big break in the API and I'm just not sure if we can justify it... I.e. the fact that you needed to change all the tests involving mask to allow them to continue to work is really not a great sign.

Is there any chance of allowing the existing code to continue working with some minimal changes, e.g. by adding `wxMaskPixelData` type and then just using it rather than `wxNativePixelData` in the tests? AFAICS it ought to be possible, but it would require implementing `Red()`, `Green()` and `Blue()` for mono format too, to return either 0xff or 0x00, but this seems doable, isn't it? And if we did it like this, we'd at least have an API which still works on all platforms, unlike in the current state of this PR.

Alternatively, and perhaps much simpler, perhaps we could keep `wxBitmap::GetMask()` unchanged, i.e. let it return 24bpp bitmap as it does now, and maybe provide `GetMonoMask()` in addition to it? This looks doable too and would avoid any compatibility breakage AFAICS (albeit at the cost of run-time pessimization).

But unfortunately I don't think we can apply this in the current state. We need

1. Portable API working under all platforms, i.e. it should be possible to avoid platform checks in our own tests.
2. As little compatibility breakage as possible.

And maybe I'm missing something, but I think we could do better on both points... Could you please check if you can do it? TIA!

wsu-cb commented:
> Is there any chance of allowing the existing code to continue working with some minimal changes, e.g. by adding `wxMaskPixelData` type and then just using it rather than `wxNativePixelData` in the tests? AFAICS it ought to be possible, but it would require implementing `Red()`, `Green()` and `Blue()` for mono format too, to return either 0xff or 0x00, but this seems doable, isn't it? And if we did it like this, we'd at least have an API which still works on all platforms, unlike in the current state of this PR.
> 

I would expect that the interface I provided in wxMSW `wxNativeMonoPixelData` (or `wxMaskPixelData` if you prefer) could be made portable to all platforms.  However, I don't think it's possible to implement `wxMaskPixelData` for a particular platform without knowing that platform's actual mask data format, is it?  I only implemented Windows because that's the only platform I can test.  Do all of the non-Windows platforms use the same format for their masks as they use for their `wxNativePixelData`?

For semantic reasons, I don't like the mono format having `Red()`, `Green()` and `Blue()`.  However, if they are required for backwards compatibility reasons, I agree that they sound easy enough to implement.

> Alternatively, and perhaps much simpler, perhaps we could keep `wxBitmap::GetMask()` unchanged, i.e. let it return 24bpp bitmap as it does now, and maybe provide `GetMonoMask()` in addition to it? This looks doable too and would avoid any compatibility breakage AFAICS (albeit at the cost of run-time pessimization).
> 

In test `BitmapTestCase::SubBitmapNonAlphaWithMask`, I use platform checks to get the right `wxPixelData` flavor for manipulating `bmpMask`.  Note that `bmpMask` is NOT the output of `wxBitmap::GetMask()`; it is directly created as a 1bpp `wxBitmap`.  Therefore, I don't think adding `GetMonoMask()` is a viable alternative to `wxMaskPixelData`.

> But unfortunately I don't think we can apply this in the current state. We need
> 
> 1. Portable API working under all platforms, i.e. it should be possible to avoid platform checks in our own tests.
> 2. As little compatibility breakage as possible.
> 
> And maybe I'm missing something, but I think we could do better on both points... Could you please check if you can do it? TIA!

wsu-cb commented on include/wx/rawbmp.h:
```diff
@@ -153,6 +153,24 @@ typedef wxPixelFormat<unsigned char, 24, 0, 1, 2> wxImagePixelFormat;
+        // doesn't cover the case of wxImage which stores alpha separately)
+        enum { HasAlpha = false };
+    };
+    typedef wxPixelFormat<void, 1, -1, -1, -1, -1, bool> wxNativeMonoPixelFormat;
```
Does this mean `wxNativeMonoPixelData` should change to `wxMonoPixelData` as well?

vadz commented:
I think preserving `GetMask()` as it is, i.e. returning 24bpp bitmap from it, and adding `GetMonoMask()` would help a lot because this would continue the existing code using `GetMask()` to work unchanged. The fact that we'd need platform checks in the tests code is not ideal, but it's not that bad neither -- while breaking all the code using `GetMask()` really is.

Maybe my two goals should have been renumbered in the opposite order:

1. Minimize or, ideally, avoid compatibility breakage.
2. Provide portable API for working with monochrome masks.

wsu-cb commented:
> I think preserving `GetMask()` as it is, i.e. returning 24bpp bitmap from it, and adding `GetMonoMask()` would help a lot because this would continue the existing code using `GetMask()` to work unchanged. The fact that we'd need platform checks in the tests code is not ideal, but it's not that bad neither -- while breaking all the code using `GetMask()` really is.
> 
> Maybe my two goals should have been renumbered in the opposite order:
> 
> 1. Minimize or, ideally, avoid compatibility breakage.
> 2. Provide portable API for working with monochrome masks.

I think we need to back up a step.  As I stated in https://github.com/wxWidgets/wxWidgets/pull/2032#issuecomment-683206974, `GetMask()` has "always" returned a mono bitmap.  However, `wxNativePixelData` only works for 24/32bpp bitmaps.  Branch master works despite this because `wxBitmap::GetRawData()` always converts mono to the 24bpp that `wxNativePixelData` uses.  (`wxPixelData` is basically a convenience wrapper for `GetRawData()`.)

Currently, `wxBitmap::GetRawData(wxPixelDataBase& data, int bpp)` only uses `bpp` to validate that it is returning what the caller expects, and fails otherwise.  Maybe what you are actually suggesting is that `GetRawData()` use the `bpp` argument to return mono data when `bpp == 1`, and RGB data when `bpp == 24`?  I think that would allow existing code using `wxNativePixelData` to work unmodified using its existing mechanism of converting between mono and RGB data, and still enable adding `wxMonoPixelData` that works with actual mono data.  However, someone else would have to implement wxMonoPixelData on non-wxMSW platforms; I can only work on wxMSW.

vadz commented:
Yes, you're right, I've misdiagnosed the level of the problem. It seems we need not a special `GetMonoMask()` but rather a special `GetRawMonoData()` -- this is more or less equivalent to adding `bpp` argument to `GetRawData()` as you propose but more compatible/less disruptive and also easier to document as being implemented only under MSW.

So, finally, I think the simplest backwards-compatible solution would be to add this function and implement it for MSW for now and just return `false` from it elsewhere. Can we do this?

Thanks again!

wsu-cb commented:
> Yes, you're right, I've misdiagnosed the level of the problem. It seems we need not a special `GetMonoMask()` but rather a special `GetRawMonoData()` -- this is more or less equivalent to adding `bpp` argument to `GetRawData()` as you propose but more compatible/less disruptive and also easier to document as being implemented only under MSW.
> 
> So, finally, I think the simplest backwards-compatible solution would be to add this function and implement it for MSW for now and just return `false` from it elsewhere. Can we do this?
> 
> Thanks again!

The `bpp` argument is already in `GetRawData()` in branch master.  `GetRawData()` uses it to verify that it is returning what the caller wants, and report an error if what `GetRawData()` knows how to return is not what the caller wants.  I am suggesting that `GetRawData()` be extended to know how to return more `bpp` formats.  I don't think this is much of a compatibility problem since it is just changing what used to be a rejected request into a correctly handled request.

vadz commented:
> The `bpp` argument is already in `GetRawData()` in branch master. 

Oops, you're right. I've trusted the comment in `wx/bitmap.h`, but it's out of date.

> `GetRawData()` uses it to verify that it is returning what the caller wants, and report an error if what `GetRawData()` knows how to return is not what the caller wants. I am suggesting that `GetRawData()` be extended to know how to return more `bpp` formats. I don't think this is much of a compatibility problem since it is just changing what used to be a rejected request into a correctly handled request.

Yes, you're absolutely correct, if calling `GetRawData(data, 1)` now is an error, there is no compatibility problem with making it return something useful, of course.

But I still don't understand something here: if compatibility is preserved, why do we need all the `#ifndef __WXMSW__` in the tests? I'd understand if we had blocks inside `#ifdef __WXMSW__` because of some functionality implemented in this port only, but shouldn't the existing tests continue to work if nothing changes for them?

wsu-cb commented:
> Yes, you're absolutely correct, if calling `GetRawData(data, 1)` now is an error, there is no compatibility problem with making it return something useful, of course.
> 
> But I still don't understand something here: if compatibility is preserved, why do we need all the `#ifndef __WXMSW__` in the tests? I'd understand if we had blocks inside `#ifdef __WXMSW__` because of some functionality implemented in this port only, but shouldn't the existing tests continue to work if nothing changes for them?

The code I have written so far still follows the branch master practice of only accepting a single `bpp` value  In other words, if wxMSW uses `GetRawData(mono_bmp, 24)` (as `wxNativePixelData` does), then `GetRawData()` reports an errror because `bpp != 1`.  I am **SUGGESTING** that `GetRawData(..., bpp)` allow both `bpp == 1` and `bpp == 24` to solve the backward compatibility problem while still enabling the addition of `wxMonoPixelData`.  At that point, the tests would indeed use `#ifdef __WXMSW__` to exercise the wxMSW-specific `wxMonoPixelData` tests, but have no `#ifndef __WXMSW__` blocks to skip `wxNativePixelData` tests.

Just to be clear:  I haven't tried to implement my suggestion; this might be harder than it sounds.  I wanted to get your agreement before going to the effort of trying.

vadz commented:
> The code I have written so far still follows the branch master practice of only accepting a single `bpp` value. In other words, if wxMSW uses `GetRawData(mono_bmp, 24)` (as `wxNativePixelData` does), then `GetRawData()` reports an errror because `bpp != 1`. I am **SUGGESTING** that `GetRawData(..., bpp)` allow both `bpp == 1` and `bpp == 24` to solve the backward compatibility problem while still enabling the addition of `wxMonoPixelData`. At that point, the tests would indeed use `#ifdef __WXMSW__` to exercise the wxMSW-specific `wxMonoPixelData` tests, but have no `#ifndef __WXMSW__` blocks to skip `wxNativePixelData` tests.
> 
> Just to be clear: I haven't tried to implement my suggestion; this might be harder than it sounds. I wanted to get your agreement before going to the effort of trying.

Understood, thanks! I do think this would be the best way of addressing the problem, as we've already discussed, this seems perfectly compatible and it should allow working with mono bitmaps, as expected with masks, and so should address your initial problem AFAICS.

So I definitely agree to doing it like this, thanks again!

wsu-cb committed:
rename wxNativeMonoPixel* to wxMonoPixel*

wsu-cb committed:
make wxBitmap::GetRawData() return caller's desired bpp

Make GetRawData() able to return 1bpp data if and only if caller
asks for it.  This allows working with monochrome bitmaps as
needed by masks without breaking compatibility with existing
code.

wsu-cb committed:
get around arith-conversion warnings-as-errors

vadz commented on src/msw/dib.cpp:
```diff
@@ -52,17 +54,19 @@
 // private functions
 // ----------------------------------------------------------------------------
 
+namespace {
```
```suggestion
namespace
{
```

wsu-cb replied:
done

vadz commented on src/msw/dib.cpp:
```diff
@@ -71,6 +75,24 @@ static inline bool GetDIBSection(HBITMAP hbmp, DIBSECTION *ds)
+    int bit = 7 - (index & 7);
+    unsigned char mask = 1 << bit;
+    return (*byte & mask) != 0;
+}
```
```suggestion
}

```

wsu-cb replied:
done

vadz commented on src/msw/dib.cpp:
```diff
@@ -654,12 +748,24 @@ bool wxDIB::Create(const wxImage& image, PixelFormat pf)
+            {
+                for ( int x = 0; x < w; x++ )
+                {
+                    wxASSERT(src[0] == 0 || src[0] == 1);
```
I think it's a bad idea to have these checks inside the loop, `wxASSERT` is present in production builds too (it's just inactive there by default) and it just doesn't seem right to have it inside an inner loop.

wsu-cb replied:
done

vadz reviewed (approved):
This looks good to me, thanks a lot for all your work on this!

I could already merge this, but I'd like to ask first: are you ok with squash-merging all these changes as a single commit or do you think any individual commits here are worth preserving? I definitely don't want to keep all 30+ commits in this PR as many of them are just "fix ups", and for me it would be simplest to just squash them all together, but if you'd like to preserve more history, please feel free to interactively rebase them and force push.

Also, can you think of a concise way to describe these changes for the change log entry?

Thanks again!

wsu-cb committed:
adjust white space

wsu-cb committed:
performance:  remove wxASSERT from frequently-executed loop

wsu-cb commented:
> I could already merge this, but I'd like to ask first: are you ok with squash-merging all these changes as a single commit or do you think any individual commits here are worth preserving? I definitely don't want to keep all 30+ commits in this PR as many of them are just "fix ups", and for me it would be simplest to just squash them all together, but if you'd like to preserve more history, please feel free to interactively rebase them and force push.

I have no problem with squashing this.


> Also, can you think of a concise way to describe these changes for the change log entry?

How concise?  Is the following too much?

wxMask expects monochrome wxBitmap, so improve support for them.
1)  Enable load and save of monochrome .bmp as monochrome wxBitmap
2)  Provide wxMonoPixelData explicit instantiation of wxPixelData
3)  Enable converting wxImage (with no mask and no alpha) to monochrome wxBitmap
Note:  only implemented in wxMSW


> Thanks again!

You're welcome

vadz commented:
@wsu-cb It looks like this has broken the appearance of disabled `wxBitmapButton`s, see the [detail here](https://trac.wxwidgets.org/ticket/18974). If you could please check what happened here, it would be great, TIA!
